### PR TITLE
Update string recognition:

### PR DESCRIPTION
**Pull request by dhardy**
Allow multi-line strings even without escaping the end of line
(rustc allows this, not sure whether it's intended).
Support raw strings: r"", r#""# and r##""## (generic version is likely not
possible).

### CODE REVIEW

**Comment by eddyb** [comment]
Generic version is possible, you can use back-references and non-greedy matches in regexp, IIRC.

File: src/etc/kate/rust.xml [hunk 1]
```diff
@@ -249,14 +252,36 @@
 			<DetectChar char="=" attribute="Normal Text" context="#pop"/>
 			<DetectChar char="&lt;" attribute="Normal Text" context="#pop"/>
 		</context>
-		<context attribute="String" lineEndContext="#pop" name="String">
-			<LineContinue attribute="String" context="#stay"/>
-			<DetectChar char="\" attribute="CharEscape" context="CharEscape"/>
-			<DetectChar attribute="String" context="#pop" char="&quot;"/>
+                <!-- It seems that rustc allows strings to extend over multiple lines, and the
+                only thing a backshash at end-of-line does is remove the whitespace.
```
Comment (huonw) [inline src/etc/kate/rust.xml]: This is the behaviour, yes.

File: src/etc/kate/rust.xml [hunk 2]
```diff
@@ -249,14 +252,36 @@
...
+                The language reference doesn't clarify. Old highlighting rule:
+                <context attribute="String" lineEndContext="#pop" name="String">
+                        <LineContinue attribute="String" context="#stay"/>
+                        <DetectChar char="\" attribute="CharEscape" context="CharEscape"/>
+                        <DetectChar attribute="String" context="#pop" char="&quot;"/>
+                </context>
+                -->
```
Comment (eddyb) [inline src/etc/kate/rust.xml]: It's intended, don't trust the reference. Shouldn't keep the old code around.

File: src/etc/kate/rust.xml [hunk 3]
```diff
@@ -249,14 +252,36 @@
...
+                <!-- These rules are't complete: they won't match r###"abc"### -->
+                <context attribute="String" lineEndContext="#stay" name="RawHashed1">
+                        <Detect2Chars attribute="String" context="StringEndGuard" char="&quot;" char1="#"/>
+                </context>
+                <context attribute="String" lineEndContext="#stay" name="RawHashed2">
+                        <StringDetect attribute="String" context="StringEndGuard" String="&quot;##"/>
+                </context>
+                <context name="StringEndGuard" attribute="Error" fallthrough="true" fallthroughContext="#pop#pop">
+                  <!-- Need a word break. This rule is probably not quite accurate. -->
+                  <RegExpr String="[#a-zA-Z0-9_&quot;&apos;]" attribute="Error"/>
```
Comment (eddyb) [inline src/etc/kate/rust.xml]: What does this achieve, exactly?

**Comment by dhardy** [comment]
The Rust reference isn't very clear on line-endings within strings. I'll update the comment.

File: src/etc/kate/rust.xml (hunk 3 above)
Comment (dhardy) [inline src/etc/kate/rust.xml]: String guard puts error highlighting on things like an extra # or " or text following string close.

**Comment by dhardy** [comment]
eddyb, care to figure out the generic version? Only way I can see is with a single regex matching the whole string, and only then if Kate supports that correctly. I can test for you.

File: src/etc/kate/rust.xml (hunk 3 above)
Comment (eddyb) [inline src/etc/kate/rust.xml]: That's unnecessary, they're both valid:
```rust
macro_rules! expr { ($x:expr) => ($x) }
// *sigh* when are we going to fix this bug with tt's?
macro_rules! foo { ($x:tt #"foo"bar ) => (expr!($x))}
fn main() {
    println!("Hello, world! {}", foo!(r#"bar"##"foo"bar))
}
```
Interestingly, github gets it right, but playpen's [pygments highlighting](https://play.rust-lang.org/?code=macro_rules%21%20expr%20%7B%20%28%24x%3Aexpr%29%20%3D%3E%20%28%24x%29%20%7D%0A//%20*sigh*%20when%20are%20we%20going%20to%20fix%20this%20bug%20with%20tt%27s%3F%0Amacro_rules%21%20foo%20%7B%20%28%24x%3Att%20%23%22foo%22bar%20%29%20%3D%3E%20%28expr%21%28%24x%29%29%7D%0Afn%20main%28%29%20%7B%0A%20%20%20%20println%21%28%22Hello%2C%20world%21%20%7B%7D%22%2C%20foo%21%28r%23%22bar%22%23%23%22foo%22bar%29%29%0A%7D) doesn't.

**Comment by dhardy** [comment]
A single-regexp version might be possible — except that Kate won't match regexps over multiple lines. So probably better not to do that.

**Comment by eddyb** [comment]
Could you please set your  github email address in your git config and squash the commits into a single, new one?

File: src/etc/kate/rust.xml [hunk 4]
```diff
@@ -249,14 +252,30 @@
...
+                <context attribute="String" lineEndContext="#stay" name="RawHashed1">
+                        <Detect2Chars attribute="String" context="StringEndGuard" char="&quot;" char1="#"/>
+                </context>
+                <context attribute="String" lineEndContext="#stay" name="RawHashed2">
+                        <StringDetect attribute="String" context="StringEndGuard" String="&quot;##"/>
+                </context>
+                <context name="StringEndGuard" attribute="Error" fallthrough="true" fallthroughContext="#pop#pop">
+                  <!-- Need a word break. This rule highlights (some) wrong characters following
+                    the end of a string. -->
+                  <RegExpr String="[#a-zA-Z0-9_&quot;&apos;]" attribute="Error"/>
```
Comment (eddyb) [inline src/etc/kate/rust.xml]: That's unnecessary, they're all valid:
```rust
macro_rules! expr { ($x:expr) => ($x) }
// *sigh* when are we going to fix this bug with tt's?
macro_rules! foo { ($x:tt #"foo"bar ) => (expr!($x))}
fn main() {
    println!("Hello, world! {}", foo!(r#"bar"##"foo"bar))
}
```
Interestingly, github gets it right, but playpen's [pygments highlighting](https://play.rust-lang.org/?code=macro_rules%21%20expr%20%7B%20%28%24x%3Aexpr%29%20%3D%3E%20%28%24x%29%20%7D%0A//%20*sigh*%20when%20are%20we%20going%20to%20fix%20this%20bug%20with%20tt%27s%3F%0Amacro_rules%21%20foo%20%7B%20%28%24x%3Att%20%23%22foo%22bar%20%29%20%3D%3E%20%28expr%21%28%24x%29%29%7D%0Afn%20main%28%29%20%7B%0A%20%20%20%20println%21%28%22Hello%2C%20world%21%20%7B%7D%22%2C%20foo%21%28r%23%22bar%22%23%23%22foo%22bar%29%29%0A%7D) doesn't.

**Comment by eddyb** [comment]
@dhardy That seems like a serious limitation - and I can't find anything in `language.dtd` that would suggest there's any way to set `RegExpr` flags (I found some "dynamic arguments" but I'm not sure how it works just yet).

File: src/etc/kate/rust.xml (hunk 4 above)
Comment (dhardy) [inline src/etc/kate/rust.xml]: What exactly are the rules? Don't follow.

**Comment by dhardy** [comment]
Yeah, I don't think Kate's reg-exprs are quite fully fledged.

File: src/etc/kate/rust.xml (hunk 4 above)
Comment (eddyb) [inline src/etc/kate/rust.xml]: Nothing is an error after a string ends (which includes the right number of trailing `#` for raw strings).

**Comment by dhardy** [comment]
Is this waiting on something?

**Comment by eddyb** [comment]
I don't want to merge a partial implementation. I couldn't find more information on the dynamic arguments, but they could handle this. If not, at least restrict to one `#`, two seems quite arbitrary.
Oh and this moved to [rust-lang/kate-config](https://github.com/rust-lang/kate-config).

**Comment by dhardy** [comment]
(1) I'm moderately sure that a generic implementation isn't possible, (2) any implementation will therefore have an arbitrary restriction [I picked two `#` symbols since there's an example using that in the docs and it seems it would cover 99%+ of uses]. Is that not okay?

But awesome that the config has moved to its own repo; will make a new pull-request.

**Pull request closed without merging**